File: Problems/Dynamic-programming/Longest-Increasing-Subsequence/Longest-Increasing-Subsequence.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// Using tabulation - Time: O(n^2)

class Solution {
public:
    int lengthOfLIS(std::vector<int> &nums)
    {
        if (nums.empty())
            return 0;

        int n = nums.size();
        std::vector<int> dp(n, 1);
        
        for (int i = 1; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (nums[j] < nums[i])
                    dp[i] = std::max(dp[i], dp[j] + 1);
            }
        }

        return *std::max_element(dp.begin(), dp.end());
    }
};

int main()
{
    std::vector<int> nums = {10, 9, 2, 5, 3, 7, 101, 18};

    std::cout << "Input: nums = [";
    for (int i = 0; i < nums.size(); i++)
    {
        std::cout << nums[i];
        if (i < nums.size() - 1)
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;  
    
    Solution sol;
    int result = sol.lengthOfLIS(nums);
    
    std::cout << "Output: " << result << std::endl;
    
    return 0;
}

File: Problems/Dynamic-programming/Longest-Increasing-Subsequence/Longest-Increasing-Subsequence.cs
using System;

// Using tabulation - Time: O(n^2)

public class Solution
{
    public int LengthOfLIS(int[] nums)
    {
        if (nums.Length == 0) return 0;

        int n = nums.Length;
        int[] dp = new int[n];
        
        for (int i = 0; i < n; i++)
        {
            dp[i] = 1; 
        }
        
        for (int i = 1; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (nums[j] < nums[i]) dp[i] = Math.Max(dp[i], dp[j] + 1);
            }
        }
        
        return dp.Max();
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[] nums = { 31, -41, 59, 26, -53, 58, 97, -93, -23, 84 };

        Console.WriteLine("Input: nums = [" + string.Join(", ", nums) + "]");

        Solution sol = new Solution();
        int result = sol.LengthOfLIS(nums);
    
        Console.WriteLine("Output: " + result);
    }
}
